### PR TITLE
Fix local Document APIs

### PR DESCRIPTION
**Pull request by rhyshort**
Change getLocalDocument to throw a DocumentNotFoundException when an
ExecutionException occurs and only log error conditions at the site
they occur, not when an ExecutionException gets thrown.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -2,6 +2,7 @@
 # 0.11.0 (Unreleased)
 
 - [FIX] Using MongoDB query as the "gold" standard, query support for `NOT` has been fixed to return result sets correctly (as in MongoDB query).  Previously, the result set from a query like `{ "pet": { "$not" { "$eq": "dog" } } }` would include a document like `{ "pet" : [ "cat", "dog" ], ... }` because the array contains an object that isn't `dog`.  The new behavior is that `$not` now inverts the result set, so this document will no longer be included because it has an array element that matches `dog`.
+- [BREAKING CHNAGE] Chnaged getLocalDocument API to throw DocumentNotFoundException when a document is not found instead of returning null and logging ExecutionException at Severe/Error level.
```
Comment (tomblench) [inline CHANGES.md]: Type CHNAGE → CHANGE, Chnaged → Changed

**Comment by rhyshort** [comment]
After discussions with @tomblench it was decided that while a breaking change was being made to the local document ids it would be a better idea to also perform changes to remove the illusion that local documents have revisions, in addition to the correction of the logging messages for getLocalDocument

File: sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java [hunk 2]
```diff
@@ -107,6 +107,7 @@ public BasicDatastore(String dir, String name) throws SQLException, IOException
         queue.updateSchema(DatastoreConstants.getSchemaVersion3(), 3);
         queue.updateSchema(DatastoreConstants.getSchemaVersion4(), 4);
         queue.updateSchema(DatastoreConstants.getSchemaVersion5(), 5);
+        queue.updateSchema(DatastoreConstants.getSchemaVersion6(), 6);
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java]: Remove this now we're not doing schema upgrades

File: sync-core/src/main/java/com/cloudant/sync/datastore/DatastoreConstants.java [hunk 3]
```diff
@@ -110,4 +110,36 @@
         };
     };
 
+//    Remember the format of all indexes and triggers associated with table X. This information will be needed in step 7 below. One way to do this is to run a query like the following: SELECT type, sql FROM sqlite_master WHERE tbl_name='X'.
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/datastore/DatastoreConstants.java]: Looks like this comment came from some sort of SQLite guide. Can we reformat it for line length and cite where it comes from? Most importantly we should make clear what this schema upgrade is doing: effectively dropping the unused revid column from localdocs

File: sync-core/src/main/java/com/cloudant/sync/datastore/LocalDocument.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package com.cloudant.sync.datastore;
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/datastore/LocalDocument.java]: Copyright headers

File: sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java (hunk 2 above)
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java]: Ignore above comment

File: CHANGES.md [hunk 5]
```diff
@@ -2,6 +2,7 @@
 # 0.11.0 (Unreleased)
 
 - [FIX] Using MongoDB query as the "gold" standard, query support for `NOT` has been fixed to return result sets correctly (as in MongoDB query).  Previously, the result set from a query like `{ "pet": { "$not" { "$eq": "dog" } } }` would include a document like `{ "pet" : [ "cat", "dog" ], ... }` because the array contains an object that isn't `dog`.  The new behavior is that `$not` now inverts the result set, so this document will no longer be included because it has an array element that matches `dog`.
+- [BREAKING CHANGE] Changed local document APIs, removed createLocalDocument and updateLocalDocument, replaced with insertLocalDocument. getLocalDocument return type changed to LocalDocument see commit for full details.
```
Comment (tomblench) [inline CHANGES.md]: Break this up into a few more sentences, eg

Changed local document APIs. createLocalDocument and updateLocalDocument have been removed and replaced by insertLocalDocument. The return type of getLocalDocument has been changed to LocalDocument.

Also see commit for full details is a bit vague - maybe refer to the specific class etc which has changed?

File: sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java (hunk 2 above)
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java]: I think you will also need some mechanism to stop an old version of our library opening a newer database - see how CDTDatastore does this.

File: sync-core/src/main/java/com/cloudant/sync/datastore/LocalDocument.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * Created by Rhys Short on 23/02/15.
+ */
+public class LocalDocument {
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/datastore/LocalDocument.java]: Some documentation would be nice

File: sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java [hunk 7]
```diff
@@ -104,9 +104,15 @@ public BasicDatastore(String dir, String name) throws SQLException, IOException
         this.extensionsDir = FilenameUtils.concat(this.datastoreDir, "extensions");
         final String dbFilename = FilenameUtils.concat(this.datastoreDir, DB_FILE_NAME);
         queue = new SQLDatabaseQueue(dbFilename);
+        int dbVersion = queue.getVersion();
+        if(dbVersion >= 100){
+            throw new DatastoreException(String.format("Database version is higher than the version supported " +
+                    "by this library, current version %d , highest supported version %d",dbVersion, 199));
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java]: Surely the highest supported version is 99?

File: sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java [hunk 8]
```diff
@@ -61,6 +66,22 @@ public Object call() throws Exception {
 
     }
 
+    public int getVersion() throws SQLException {
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java]: javadoc? Something like `Returns the schema version of the underlying SQL database`

**Comment by tomblench** [comment]
OK I think this in pretty good shape now. The PR summary probably needs to change to reflect what's been done here. How about "Fix local document APIs" as a headline and then the details in a squashed commit message about what's changed, including schema changes etc.

**Comment by rhyshort** [comment]
Yeah sounds, good to me, I'll make those changes

File: sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java [hunk 9]
```diff
@@ -680,48 +667,37 @@ private void validateDBBody(DocumentBody body) {
...
-                        }
+                public LocalDocument call(SQLDatabase db) throws Exception {
+                    ContentValues values = new ContentValues();
+                    values.put("docid", docId);
+                    values.put("json", body.asBytes());
+
+                    long rowId = db.insertWithOnConflict("localdocs", values, SQLDatabase
+                            .CONFLICT_REPLACE);
+                    if (rowId < 0) {
+                        throw new DocumentException("Failed to insert local document");
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java]: These log messages and exception messages here and below are a bit varied: "failed/cannot", "create/insert" - they should be consistent.

File: sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java [hunk 10]
```diff
@@ -62,6 +67,27 @@ public Object call() throws Exception {
...
+        try {
+            return queue.submit(new Callable<Integer>() {
+                 @Override
+                 public Integer call() throws Exception {
+                     return db.getVersion();
+                 }
+             }).get();
+        } catch (InterruptedException e) {
+            logger.log(Level.SEVERE,"Failed to get database version",e);
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java]: Log for both catch blocks or neither.

**Comment by tomblench** [comment]
OK +1 after you've fixed up those last minor comments

**Pull request merged**